src/LinkDotNet.NCronJob/Execution/JobExecutor.cs

```diff
@@ -33,32 +44,36 @@ public void RunJob(RegistryEntry run, CancellationToken stoppingToken)
             return;
         }
 
-        ExecuteJob(run, job, scope, stoppingToken);
+        await ExecuteJob(run, job, scope, stopToken);
     }
 
-    public void Dispose() => isDisposed = true;
+    public void Dispose()
+    {
+        shutdown?.Dispose();
+        isDisposed = true;
+    }
 
-    private void ExecuteJob(RegistryEntry run, IJob job, IServiceScope serviceScope, CancellationToken stoppingToken)
+    private async Task ExecuteJob(RegistryEntry run, IJob job, IServiceScope serviceScope, CancellationToken stoppingToken)
     {
         try
         {
             LogRunningJob(run.Type);
 
-            // We don't want to await jobs explicitly because that
-            // could interfere with other job runs
-            job.RunAsync(run.Context, stoppingToken)
-                .ContinueWith(
-                    task => AfterJobCompletionTask(task.Exception),
-                    TaskScheduler.Default)
+            // Job Runs are executed in parallel by default
+            await job.RunAsync(run.Context, stoppingToken)
                 .ConfigureAwait(false);
+
+            stoppingToken.ThrowIfCancellationRequested();
+
+            await AfterJobCompletionTask(null, stoppingToken);
         }
         catch (Exception exc) when (exc is not OperationCanceledException or AggregateException)
         {
             // This part is only reached if the synchronous part of the job throws an exception
-            AfterJobCompletionTask(exc);
+            await AfterJobCompletionTask(exc, default);
         }
-
-        void AfterJobCompletionTask(Exception? exc)
+        // This needs to be async otherwise it can deadlock or try to use the disposed scope, maybe it needs to create its own serviceScope
```

Comment:
In my mental model, the completion (and therefore the notification server handlers) are somewhat a part of the Job itself and, therefore, share the same scope.

Reply:
If a Job is cancelled then the subsequent AfterJobCompletionTask would not get triggered within the Try block. Are you suggesting to remove the call from within the catch? I left that there because that's how you had it, but I don't pass the CancellationToken to avoid the notifications from being cancelled. I agree this needs a change.